be/src/storage/lake/pk_index_loader.cpp

```diff
@@ -0,0 +1,208 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/lake/pk_index_loader.h"
+
+#include <fmt/format.h>
+
+#include <chrono>
+
+#include "common/constexpr.h"
+#include "exec/workgroup/scan_executor.h"
+#include "exec/workgroup/work_group.h"
+#include "storage/chunk_helper.h"
+#include "storage/lake/update_manager.h"
+#include "storage/olap_common.h"
+#include "storage/rowset/segment_options.h"
+#include "util/defer_op.h"
+
+namespace starrocks::lake {
+
+class PkSegmentScanTask {
+public:
+    PkSegmentScanTask(Tablet* tablet, uint32_t rowset_id, uint32_t seg_id, const std::string& seg_name,
+                      const Schema& schema, int64_t version, const MetaFileBuilder* builder)
+            : _tablet(tablet),
+              _rowset_id(rowset_id),
+              _seg_id(seg_id),
+              _seg_name(std::move(seg_name)),
+              _schema(std::move(schema)),
+              _version(version),
+              _builder(builder) {}
+
+    void run() {
+        auto st = Status::OK();
+        DeferOp defer([&]() {
+            if (st.is_end_of_file()) {
+                st = Status::OK();
+            }
+            ExecEnv::GetInstance()->lake_pk_index_loader()->finish_subtask(_tablet->id(), st);
+        });
+        size_t footer_size_hint = 16 * 1024;
+        auto seg = _tablet->load_segment(_seg_name, _seg_id, &footer_size_hint, false, false);
+        if (!seg.ok()) {
+            st = seg.status();
+            return;
+        }
+        SegmentReadOptions seg_options;
+        auto fs = FileSystem::CreateSharedFromString(_tablet->root_location());
+        if (!fs.ok()) {
+            st = fs.status();
+            return;
+        }
+        OlapReaderStatistics stats;
+        seg_options.fs = *fs;
+        seg_options.stats = &stats;
+        seg_options.is_primary_keys = true;
+        seg_options.delvec_loader = std::make_shared<LakeDelvecLoader>(_tablet->update_mgr(), _builder);
+        seg_options.version = _version;
+        seg_options.tablet_id = _tablet->id();
+        seg_options.rowset_id = _rowset_id;
+        auto seg_itr = (*seg)->new_iterator(_schema, seg_options);
+        if (!seg_itr.ok()) {
+            st = seg_itr.status();
+            return;
+        }
+        while (true) {
+            auto chunk_info = std::make_shared<ChunkInfo>();
+            chunk_info->rowids.reserve(DEFAULT_CHUNK_SIZE);
+            chunk_info->chunk = ChunkHelper::new_chunk(_schema, DEFAULT_CHUNK_SIZE);
+            chunk_info->rssid = _rowset_id + _seg_id;
+            auto chunk = chunk_info->chunk.get();
+            st = (*seg_itr)->get_next(chunk, &chunk_info->rowids);
+            if (st.is_end_of_file()) {
+                break;
+            } else if (!st.ok()) {
+                return;
+            }
+            ExecEnv::GetInstance()->lake_pk_index_loader()->add_chunk(_tablet->id(), chunk_info);
```

Comment:
Is it possible that in extreme cases, the rate of consumer cannot keep up with the producer？ I mean，`add_chunk` may take lots of memory if `get_chunk` not be called timely

Reply:
In theory yes. Maybe a blocking queue can be used later if these scenario occurs.

Reply:
I agree with @smartlxh and I think this is a critical issue that should be addressed in this PR.